### DISCUSSION TITLE
Local storage on initial site render

### DISCUSSION THREAD

**Discussion by ufelgen** [neuefische/questions#401] [Web] [answered]
Hi everyone, 
I have a new local storage problem. 
I successfully implemented the local storage hook to store my data, its initial value is my dummy array:
`const [allWords, setAllWords] = useLocalStorage("allWords", words);`
I can access this (sometimes, see below) on my dynamic pages by calling the hook function again:
`const [allWords] = useLocalStorage("allWords");`
I have a form with a submit handler function that successfully adds new info to the data stored in the local storage.
HOWEVER, only upon submitting new data through the form does the whole data set appear in the local storage. Before that (on initial site render) the local storage is empty. This means that my dynamic pages, which need to access data from the local storage, are empty. 

Can someone please help me store my data in local storage on initial site render?
Thank you so much!

**Comment by ThomasKirst** [comment]
Can you please provide a link to a PR or a branch in your repository where we can have a look into it?

**Reply by ufelgen** [reply]
https://github.com/ufelgen/capstone-project/pull/9
thank you

**Comment by julialoeschel** [comment, marked as answer]
Hello Ully, 

the provided useLocalStorage hook works exactly like this. 
As an alternative, you should use [`use-local-storage-state`](https://www.npmjs.com/package/use-local-storage-state)

The usage is almost the same, **but you need to use `useLocalStorageState` instead `useLocalStorage`** and for the initial set, you it will look like this
```
 useLocalStorageState("allWords", {
        defaultValue: words
    } )

```

**Reply by ufelgen** [reply]
Hi Julia, 
thank you for your feedback. 
I copied the code you provided into my index.js, however it still won't show anything in the local storage upon initial rendering. Did I do it wrong?

**Reply by ThomasKirst** [reply]
Please check again, the provided `useLocalStorageState` hook should make this work as expected, see for reference:
https://codesandbox.io/s/todos-example-use-local-storage-state-tzbfhl

You might need to refresh the view in the local storage tab within the developer tools.

**Reply by ufelgen** [reply]
I added a comment a while ago, but it disappeared?! Sorry. 

I think I fixed it with a useEffect in the index.js: 
`  useEffect(() => {
    localStorage.setItem("allWords", JSON.stringify(allWords));
  }, []);
`

it works, can I do it like this?

**Reply by ufelgen** [reply]
IMPORTANT:

the useLocalStorage hook provided in the capstone template does NOT store the default data in the local storage on initial site render. 

writing the default data into the local storage using a useEffect works, but IS NOT RECOMMENDED (see below)

the solution is to use the use-local-storage-state library like this: 

`npm install use-local-storage-state`

in the index.js:
`import useLocalStorageState from "use-local-storage-state";`
`const [allWords, setAllWords] = useLocalStorageState("allWords", {
    defaultValue: words,
  });
`

in the dynamically rendered pages: 
`import useLocalStorageState from "use-local-storage-state";`
`const [allWords] = useLocalStorageState("allWords");`

**Comment by ufelgen** [comment]
I think I did it!!!! I added a useEffect in my index.js:

`  useEffect(() => {
    localStorage.setItem("allWords", JSON.stringify(allWords));
  }, []);`

**Reply by ThomasKirst** [reply]
Please try to avoid the `useEffect()`, the localStorage hook should provide this functionality.

**Reply by ufelgen** [reply]
sorry for the confusion, my comment was gone, now I can see it again, together with your answer. Okay, I will try and find a different way, thank you for the sandbox link
